pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -859,6 +865,21 @@ void receiveIndividualMessagesFromBatch(MessageMetadata msgMetadata, ByteBuf unc
                     .newBuilder();
                 ByteBuf singleMessagePayload = Commands.deSerializeSingleMessageInBatch(uncompressedPayload,
                     singleMessageMetadataBuilder, i, batchSize);
+
+                if (startMessageId != null
+                        && messageId.getLedgerId() == startMessageId.getLedgerId()
+                        && messageId.getEntryId() == startMessageId.getEntryId()
+                        && i <= startMessageId.getBatchIndex()) {
+                    // If we are receiving a batch message, we need to discard messages that were prior
+                    // to the startMessageId
```

Comment:
just a quick suggestion if that make sense.
1. what if we also add similar validation at `messageReceived()` for non-batch-msg also where we skip the message if `receivedMsgId < startMessageId`
2. and at [broker ](https://github.com/merlimat/pulsar/blob/fa752f4f990c7cd604bee8fea3cc5aaf33bfe291/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java?utf8=%E2%9C%93#L477-L483) we always read from `entryId=entryId-1` and duplicate entry will be anyway skipped by the client.

In this way, broker can be `BatchMessage` agnostic and it will not have any batch-msg specific logic as it treats  batch-msg as normal msg only.

Reply:
I think that checking everything in client side gets a bit more complicated for the comparison. Also we'd have to do it in 2 places one for batch messages and one for regular. Plus in C++ and Java.

Reply:
> I think that checking everything in client side gets a bit more complicated for the comparison.

Oh. Other than adding skip-condition for non-batch msg, do we have to do anything additionally? My take would be, handling duplicate msg at client (because anyway we are doing for batch-msg so, why not for normal msg) is better than adding small-hack at [broker](https://github.com/merlimat/pulsar/blob/fa752f4f990c7cd604bee8fea3cc5aaf33bfe291/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java?utf8=%E2%9C%93#L477-L483).

Reply:
Uhm, I still prefer doing it in server side :) 

What about encapsulating in the `MessageIdImpl` so that we can have a method to get the "right" message id and that method can be overriden in the `BatchMessageIdImpl` class?

Reply:
> Uhm, I still prefer doing it in server side :)

Alright.. 👍